### PR TITLE
[wip] Templates

### PR DESCRIPTION
**Pull request by bprashanth**
This is not getting checked in as is, it's for anyone who might want to pick it up before I get time to do it. Feel free to clone this and open another pr, or just use it as a guideline. 

See https://github.com/kubernetes/kubernetes/issues/23896 for context.

### CODE REVIEW

File: pkg/registry/template/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		// by combining the namespace in the context with the given prefix
+		KeyRootFunc: func(ctx api.Context) string {
+			return etcdgeneric.NamespaceKeyRootFunc(ctx, prefix)
+		},
+		// Produces a templates that etcd understands, to the resource by combining
+		// the namespace in the context with the given prefix
+		KeyFunc: func(ctx api.Context, name string) (string, error) {
+			return etcdgeneric.NamespaceKeyFunc(ctx, prefix, name)
+		},
+		// Retrieve the name field of a replication controller
```
Comment (bparees) [inline pkg/registry/template/etcd/etcd.go]: name field of a template?

File: pkg/registry/template/guestbook_template.json [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    "objects": [
+        {
+            "apiVersion": "v1",
+            "kind": "ReplicationController",
+            "metadata": {
+                "creationTimestamp": null,
+                "labels": {
+                    "name": "redis-master"
+                },
+                "name": "${MASTER_NAME}"
```
Comment (bparees) [inline pkg/registry/template/guestbook_template.json]: the various ${} substitutions in here should be $()

File: pkg/registry/template/template.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/validation/field"
+
+	"github.com/golang/glog"
+	// TODO: Fix this import
+	"k8s.io/kubernetes/pkg/util/labels"
+	"k8s.io/kubernetes/pkg/util/stringreplace"
+)
+
+var parameterExp = regexp.MustCompile(`\$\{([a-zA-Z0-9\_]+)\}`)
```
Comment (bparees) [inline pkg/registry/template/template.go]: will need to change {} to () and in general the substitution logic will need to be updated to also find $(()) and handle the quoting separately.

**Comment by bgrant0607** [comment]
cc @pwittrock @janetkuo

**Comment by pwittrock** [comment]
Picking this up in #24777

**Comment by soltysh** [comment]
@pwittrock can you close this PR in that case?

**Pull request closed without merging**